Docs/include sub components

SinisterMaya opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
<!-- Describe your changes in detail any why -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->
This one is more complex, so I split it in different commits so each one can be analysed and understood

1. **Added Summary description to some DataGrid sub components**
This one is self explanatory. If the result is satisfactory, I will go ahead and try to expand this to other component that yet lacks description.

![image](https://user-images.githubusercontent.com/29356866/236637143-6c3a69af-6d30-4b74-beb2-feeaeb9fe460.png)


2. **Fix to the Doc page header that display the list of API link.**
Changed it so it displays correctly PropertyColumn that was displaying PropertyColumn`2 (I guess because it contains two generic type reference ?). Also changed the href to use the APILink specialCaseComponent that will help differentiate MudTemplateColumn (From MudTable) from TemplateColumn (from DataGrid).

3. **Added TProperty class**
Was required so the code compile when referencing to PropertyColumn<T, TProperty>.

4. ~~**Fix a displaying bug in DocString when fetching summary Description**~~
~~The code was including newlines from summary tags. they are now replace with whitespace so we do not introduce random newline into the doc description section.~~
EDIT: This was not a bug. It helps with long description such as Validation

5. **Added SubComponent in reference to DataGrid**
For now, PropertyColumn, TemplateColumn, SelectColumn and Pager components are added as API Reference to the Datagrid Docs

![image](https://user-images.githubusercontent.com/29356866/236637111-d6aef044-8be5-43a4-b87d-da413b724caa.png)

6. **Added CellTemplate example**
Often times, when helping people on Discord about how to use DataGrid, it involves CellTemplate and how to leverage that in order to get the full potential out of DataGrid. This is a basic example of how you can use EventCallBack from a component into the cell that will expose both the item from Items and the property from which the EventCallback was invoked.

## How Has This Been Tested?
<!-- All PR's should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
All these changes target the Documentation project only. the changes into MudBlazor Components are only for the summary descriptions. I don't know of a method to test this other than navigating the documentations and validating that nothing's broken

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

iwis commented:
> I don't know of a method to test this other than navigating the documentations and validating that nothing's broken

I cannot review the changes but for testing you can use https://github.com/MudBlazor/MudBlazor/discussions/3595 as I was using it in my PRs.

SinisterMaya commented:
Added another example in DataGrid to show how it can be possible to drive data by both columns and rows. 
![image](https://github.com/MudBlazor/MudBlazor/assets/29356866/94cdd6ab-09bc-4b4c-89a7-e80e64507f05)

Commit:
Docs: Added Summary for additionnal description

Commit:
Docs: Added Recursion Linq Extension

Commit:
Docs: Added Table and DataGrid Subcomponents Api link

Commit:
Docs: Added Class pages. Adds Support and detection for referenced type in api pages

Commit:
Docs: Added more advanced datagrid examples

SinisterMaya commented:
@Mr-Technician I'd appreciate if you could review this. It might need some polish before it's up to standard, but I didn't want to go too far down the rabbit hole before getting validation from the core team. I got class pages working, added some description and extended API documentation.

One thing I wasn't able to find how to make it work was to make the ScrollSpy work with the section I added to both DocsApi and ClassApi.

https://github.com/MudBlazor/MudBlazor/assets/29356866/49a51c5b-4b25-4bc6-8e24-12267951a47b

Let me know if something is not clear. Thanks!

ScarletKuro commented:
Hi! I'd like to propose removing the usage of camelCase for some of the methods you have and use PascalCase(regardless of the method's modifier or whether it's used in a Razor file or a typical class). Thank you for your consideration!

SinisterMaya commented:
Hi!

I will review my code. I usually camelCase variable and, on some occasions, private fields. Methods I make are always PascalCase. Must be a typo from my part !

Mr-Technician reviewed (commented):
Looking good so far!

Edit: I haven't been able to look locally yet because I can't build MudBlazor due to SDK issues and don't currently have the time to dig into that, but the screenshots and video look good.

Mr-Technician commented:
> Hi! I'd like to propose removing the usage of camelCase for some of the methods you have and use PascalCase(regardless of the method's modifier or whether it's used in a Razor file or a typical class). Thank you for your consideration!

I second this.

Commit:
Docs: Change method in docs to be PascalCase

Mr-Technician commented:
@SinisterMaya Are you finished working on this PR?

SinisterMaya commented:
@Mr-Technician yes, I probably won't be able to add anything in the upcoming days since my first child will probably birth today. If anything is needed to change, I can do this in another PR. I know Snake was waiting for this PR so they could do their own PR over it.

Mr-Technician commented:
Congratulations! I will give this a more thorough review when I am able.

ScarletKuro commented:
@henon @danielchalmers @jperson2000 should we close this? It will be barely compatible after https://github.com/MudBlazor/MudBlazor/pull/8903? 

Maybe @jperson2000 can take some inspiration from this PR.

SinisterMaya commented:
Wow it's already been a year. Yeah, my kid took a lot more time out of me than I expected. I have been out of the loop for a while, but I might be able review it myself. My intention was to generate the docs for every other objects that are not straight up usable components. For instance, what is supposed to be the structure of a group definition. If the docs example does not cover every parameters, I had to open the source code to go check myself or I had to do some trial and error with intellisense to see if the property was doing what I wanted to do.

jperson2000 commented:
> @henon @danielchalmers @jperson2000 should we close this? It will be barely compatible after #8903?
> 
> Maybe @jperson2000 can take some inspiration from this PR.

Oh wow, yeah these are great ideas @SinisterMaya !  Showing nested types and types mentioning another type would be really helpful.  Even if this PR is closed, there's valuable code here worth integrating into future docs PR's and I've bookmarked this page.

henon commented:
@SinisterMaya closing. In any case, thanks, and I hope it will be revived someday.
